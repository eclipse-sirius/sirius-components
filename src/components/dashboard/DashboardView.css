/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

.dashboardview {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: min-content min-content;
  grid-row-gap: var(--layoutDimension-l);
}

.dashboardview .info {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  grid-column-gap: var(--layoutDimension-l);
  grid-row-gap: var(--layoutDimension-l);
}

.projectsinfo {
  background-color: var(--blue);
  color: var(--white);
}

.viewpointsinfo {
  background-color: var(--orange);
  color: var(--white);
}

.metamodelsinfo {
  background-color: var(--purple);
  color: var(--white);
}

.dashboardview .projects-body {
  display: grid;
  grid-auto-rows: 1fr;
  grid-template-rows: 1fr 1fr;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  grid-row-gap: var(--layoutDimension-l);
  grid-column-gap: var(--layoutDimension-l);
}
