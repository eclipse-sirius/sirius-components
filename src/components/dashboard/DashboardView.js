/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { classNames } from '../../common/classnames';
import { UNSUPPORTED_STATE } from '../../common/errors';

import { ErrorCard } from '../error/ErrorCard';
import { InfoCard } from '../info/InfoCard';
import { Loading } from '../loading/Loading';
import { ProjectSummaryCard } from '../projects/ProjectSummaryCard';

import {
  ERROR__STATE,
  LOADING__STATE,
  DASHBOARD_LOADED__STATE
} from './DashboardViewFiniteStateMachine';

import './DashboardView.css';

const propTypes = {
  stateId: PropTypes.string.isRequired
};

/**
 * The DashboardView component is used as the main component in the dashboard page.
 * It will render a bird eye view of the state of the data of the user starting
 * with the list of the projects available.
 */
export const DashboardView = ({ className, stateId, error, dashboard, ...props }) => {
  switch (stateId) {
    case LOADING__STATE:
      return renderLoadingState(className, props);
    case ERROR__STATE:
      return renderErrorState(className, error, props);
    case DASHBOARD_LOADED__STATE:
      return renderDashboardLoadedState(className, dashboard, props);
    default:
      const undefinedStateError = {
        title: `The dashboard is in an unsupported state: ${stateId}`,
        message: 'Contact your administrator to find a suitable solution',
        code: UNSUPPORTED_STATE
      };
      return renderErrorState(className, undefinedStateError, props);
  }
};
DashboardView.propTypes = propTypes;

/**
 * Renders the loading state of the dashboard.
 * @param {*} className The class name of the dashboard.
 * @param {*} props The properties of the component
 */
const renderLoadingState = (className, props) => <Loading className={className} {...props} />;

/**
 * Renders the error.
 * @param {*} className The class name of the dashboard
 * @param {*} error The error to render
 * @param {*} props The properties of the component
 */
const renderErrorState = (className, error, props) => (
  <ErrorCard className={className} {...error} {...props} />
);

const DASHBOARD_VIEW__CLASS_NAMES = 'dashboardview';
const INFO__CLASS_NAMES = 'info';
const PROJECTS_INFO__CLASS_NAMES = 'projectsinfo';
const VIEWPOINTS_INFO__CLASS_NAMES = 'viewpointsinfo';
const METAMODELS_INFO__CLASS_NAMES = 'metamodelsinfo';
const NEWPROJECT_INFO__CLASS_NAMES = 'newprojectinfo';
const PROJECTS__CLASS_NAMES = 'projects';
const PROJECTS_BODY__CLASS_NAMES = 'projects-body';

/**
 * Renders the dashboard.
 * @param {*} className The class name of the dashboard
 * @param {*} dashboard The dashboard to display
 * @param {*} props The properties of the component
 */
const renderDashboardLoadedState = (className, dashboard, props) => {
  const dashboardViewClassNames = classNames(DASHBOARD_VIEW__CLASS_NAMES, className);

  const { projectsCount, viewpointsCount, metamodelsCount } = dashboard;

  return (
    <div className={dashboardViewClassNames} {...props}>
      <div className={INFO__CLASS_NAMES}>
        <InfoCard
          className={PROJECTS_INFO__CLASS_NAMES}
          title={projectsCount.toString()}
          message={'Projects'}
        />
        <InfoCard
          className={VIEWPOINTS_INFO__CLASS_NAMES}
          title={viewpointsCount.toString()}
          message={'Viewpoints'}
        />
        <InfoCard
          className={METAMODELS_INFO__CLASS_NAMES}
          title={metamodelsCount.toString()}
          message={'Metamodels'}
        />
      </div>
      <div className={PROJECTS__CLASS_NAMES}>
        <div className={PROJECTS_BODY__CLASS_NAMES}>
          <Link to="/newproject">
            <InfoCard
              className={NEWPROJECT_INFO__CLASS_NAMES}
              title="+"
              message="New Project"
              to=""
            />
          </Link>
          {dashboard.projects.map(project => (
            <ProjectSummaryCard key={project.name} project={project} />
          ))}
        </div>
      </div>
    </div>
  );
};
