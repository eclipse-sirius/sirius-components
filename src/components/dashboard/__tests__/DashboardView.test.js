/*******************************************************************************
 * Copyright (c) 2018 Obeo and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2
 * which accompanies this distribution and is available at
 * https://www.eclipse.org/legal/epl-2.0.
 *******************************************************************************/

import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import Renderer from 'react-test-renderer';

import { DashboardView } from '../DashboardView';
import {
  ERROR__STATE,
  INITIAL__STATE,
  LOADING__STATE,
  DASHBOARD_LOADED__STATE
} from '../DashboardViewFiniteStateMachine';

describe('DashboardView', () => {
  it('renders an error card for an error', () => {
    const error = {
      title: 'An error has occurred',
      message: 'Please contact the administrator to try to find a solution',
      code: 500
    };
    const dashboard = {
      projectsCount: 0,
      viewpointsCount: 27,
      metamodelsCount: 132,
      projects: []
    };
    const dashboardComponent = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <DashboardView stateId={ERROR__STATE} error={error} dashboard={dashboard} />
      </MemoryRouter>
    );
    expect(dashboardComponent.toJSON()).toMatchSnapshot();
  });

  it('renders an error card for an unsupported state', () => {
    const dashboard = {
      projectsCount: 0,
      viewpointsCount: 27,
      metamodelsCount: 132,
      projects: []
    };
    const dashboardComponent = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <DashboardView stateId={INITIAL__STATE} error={null} dashboard={dashboard} />
      </MemoryRouter>
    );
    expect(dashboardComponent.toJSON()).toMatchSnapshot();
  });

  it('renders a loading indicator during the loading', () => {
    const dashboard = {
      projectsCount: 0,
      viewpointsCount: 27,
      metamodelsCount: 132,
      projects: []
    };
    const dashboardComponent = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <DashboardView stateId={LOADING__STATE} error={null} dashboard={dashboard} />
      </MemoryRouter>
    );
    expect(dashboardComponent.toJSON()).toMatchSnapshot();
  });

  it('renders the projects', () => {
    const projects = [{ name: 'First Project' }, { name: 'Second Project' }];
    const dashboard = {
      projectsCount: 0,
      viewpointsCount: 27,
      metamodelsCount: 132,
      projects
    };
    const dashboardComponent = Renderer.create(
      <MemoryRouter initialEntries={['/']}>
        <DashboardView stateId={DASHBOARD_LOADED__STATE} error={null} dashboard={dashboard} />
      </MemoryRouter>
    );
    expect(dashboardComponent.toJSON()).toMatchSnapshot();
  });
});
