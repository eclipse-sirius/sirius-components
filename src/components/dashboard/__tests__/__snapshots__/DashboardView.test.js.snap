// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardView renders a loading indicator during the loading 1`] = `
<div
  className={undefined}
>
  Loading
</div>
`;

exports[`DashboardView renders an error card for an error 1`] = `
<div
  className="card errorcard"
>
  <h1
    className="title-l"
  >
    An error has occurred
  </h1>
  <p>
    Please contact the administrator to try to find a solution (code: 500)
  </p>
</div>
`;

exports[`DashboardView renders an error card for an unsupported state 1`] = `
<div
  className="card errorcard"
>
  <h1
    className="title-l"
  >
    The dashboard is in an unsupported state: INITIAL__STATE
  </h1>
  <p>
    Contact your administrator to find a suitable solution (code: 1001)
  </p>
</div>
`;

exports[`DashboardView renders the projects 1`] = `
<div
  className="dashboardview"
>
  <div
    className="info"
  >
    <div
      className="card infocard projectsinfo"
    >
      <h1
        className="infocard-title"
      >
        0
      </h1>
      <p
        className="infocard-message"
      >
        Projects
      </p>
    </div>
    <div
      className="card infocard viewpointsinfo"
    >
      <h1
        className="infocard-title"
      >
        27
      </h1>
      <p
        className="infocard-message"
      >
        Viewpoints
      </p>
    </div>
    <div
      className="card infocard metamodelsinfo"
    >
      <h1
        className="infocard-title"
      >
        132
      </h1>
      <p
        className="infocard-message"
      >
        Metamodels
      </p>
    </div>
  </div>
  <div
    className="projects"
  >
    <div
      className="projects-body"
    >
      <a
        href="/newproject"
        onClick={[Function]}
      >
        <div
          className="card infocard newprojectinfo"
          to=""
        >
          <h1
            className="infocard-title"
          >
            +
          </h1>
          <p
            className="infocard-message"
          >
            New Project
          </p>
        </div>
      </a>
      <div
        className="card projectsummarycard"
      >
        <a
          href="/projects/First Project"
          onClick={[Function]}
        >
          <h2
            className="card-primarytitle title-l"
          >
            First Project
          </h2>
        </a>
      </div>
      <div
        className="card projectsummarycard"
      >
        <a
          href="/projects/Second Project"
          onClick={[Function]}
        >
          <h2
            className="card-primarytitle title-l"
          >
            Second Project
          </h2>
        </a>
      </div>
    </div>
  </div>
</div>
`;
